capsules/src/i2c_master_slave_driver.rs

```diff
@@ -222,73 +237,107 @@ impl hil::i2c::I2CHwSlaveClient for I2CMasterSlaveDriver<'_> {
 impl Driver for I2CMasterSlaveDriver<'_> {
     fn allow_readwrite(
         &self,
-        _appid: ProcessId,
+        app: ProcessId,
         allow_num: usize,
         mut slice: ReadWriteAppSlice,
     ) -> Result<ReadWriteAppSlice, (ReadWriteAppSlice, ErrorCode)> {
-        match allow_num {
-            // Pass in a buffer for transmitting a `write` to another
-            // I2C device.
-            0 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.master_tx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            // Pass in a buffer for doing a read from another I2C device.
-            1 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.master_rx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            // Pass in a buffer for handling a read issued by another I2C master.
-            2 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.slave_tx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            // Pass in a buffer for handling a write issued by another I2C master.
-            3 => {
-                self.app.map(|app| {
-                    core::mem::swap(&mut app.slave_rx_buffer, &mut slice);
-                });
-                Ok(slice)
-            }
-            _ => Err((slice, ErrorCode::NOSUPPORT)),
+        let res = self
+            .apps
+            .enter(app, |app| {
```

Comment:
Nit: this technically works, though the `ProcessId` and `Grant` region are related but different things and I fear that shadowing the `app` variable might be confusing. Feel free to not change this if not considered an issue.